### ISSUE TITLE
# shellcheck disable=SC2029 would be nicer inside the remote string

### ISSUE THREAD

**Issue by brianjmurrell** [koalaman/shellcheck#1038]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
SC2029

- My shellcheck version (`shellcheck --version` or "online"): 
ShellCheck - shell script analysis tool
version: 0.4.5
license: GNU General Public License, version 3
website: http://www.shellcheck.net

- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

LOCAL_STRING="foo"
ssh root@remote_host "ls -l
# shellcheck disable=SC2029
echo $LOCAL_STRING"
```

#### Here's what shellcheck currently says:
Line 6:
echo $LOCAL_STRING"
     ^-- SC2029: Note that, unescaped, this expands on the client side.

#### Here's what I wanted or expected to see:
It would be nice if the `# shellcheck disable=SC2029` could be put into the remote side string/script (as I have done above in my example) so that one can indicate which occurrences are to be expected and which are not.  Requiring the disable before the whole remote string may lead to instances inside the string that should be flagged which are ignored, in the case of a many-line remote script.

**Comment by brother** [comment]
Is SC2029 just an example here?
If you put the disable before the SSH command all will be well. I think I understand the thing you are trying to put forward if you had some other warning to silence but I don't see the real need for it. The shellcheck command will never execute over the code on the other side of the wire and having escapes in that part is not very helpful really.

**Comment by brianjmurrell** [comment]
> Is SC2029 just an example here?

No, it's specific to "remote code" being executed over an _ssh_ for example.

> If you put the disable before the SSH command all will be well.

Right.  But covers every instance of variables in the _remote code_.  I.e. it's an all-or-none.

> I think I understand the thing you are trying to put forward if you had some other warning to silence but I don't see the real need for it. The shellcheck command will never execute over the code on the other side of the wire and having escapes in that part is not very helpful really.

It does seem to execute at least for SC2029 though.  I get a warning for unescaped variables.  Taking a closer look though, I do notice that I only get the warning for the first one in the remote code, no matter how many there are so maybe my suggestion is not all that helpful in any case.

I guess ultimately having per variable warnings in remote code where you disable the ones you ack that should be unescaped and get warnings about those you have not disabled but not escaped would be useful to catch inadvertent unescaping.

**Comment by Nightfirecat** [comment]
An alternative solution might be to compose the remote command string before executing `ssh` and disabling SC2029 for the expansion of the remote command string:

```
#!/bin/bash

LOCAL_STRING="foo"
REMOTE_COMMAND="ls -l
echo $LOCAL_STRING"
# shellcheck disable=SC2029
ssh root@remote_host "$REMOTE_COMMAND"
```

**Comment by brianjmurrell** [comment]
@Nightfirecat Isn't that effectively the same as just:

```
#!/bin/bash

LOCAL_STRING="foo"
# shellcheck disable=SC2029
ssh root@remote_host "ls -l
echo $LOCAL_STRING"
```

Where you don't have SC2029 helping you flag instances of variables in the remote command that need to be escaped -- meaning all of those which you wouldn't have applied a "disable=SC2029" to individually?

What I am getting at here is that it would be nice to apply that disable more granular than the whole remote script to help catch errors.

**Comment by koalaman** [comment]
I definitely see your point, and this would indeed be useful.

However, for the immediate shell, the script passed to ssh is just a string, and ShellCheck intentionally makes no attempt to parse strings passed to `ssh`, `eval`, `su`, `sh -c` or any other way of evaluating strings as shell code.

It would be useful to not emit this for any variable expanded with escaping though, such as `${var@Q}`. This is a strong signal that the local expansion was intentional.

**Comment by brianjmurrell** [comment]
> ShellCheck intentionally makes no attempt to parse strings passed to ssh, eval, su, sh -c or any other way of evaluating strings as shell code

Is it too complicated in the context of shellcheck to do that?  If shellcheck sees:
```
$ ssh remote_host "#!/bin/bash
ls -l \$somefile"
```

It would indeed be useful to have shellcheck run over the remote script.  But not having any idea how shellcheck is implemented, perhaps I am just not grasping how much more difficult that is than running over the local script.

**Comment by koalaman** [comment]
It's easy to run it over static strings, and this is basically how backticks already work (unlike `$()` they are defined in terms of a string span and not by a token like `)`, which is why they nest so poorly).

The problem is when the string is not static, such as `name="var$RANDOM"; eval "$name=value"` or `functions=$(wget -O - https://example.com/myfunctions.sh); ssh host "$functions; do_stuff \$bar";` . Like a bad IRC/stackoverflow question, it's hard to give useful and accurate feedback on a script when you're missing pieces of it. 

Since the dynamic case is undecidable, ShellCheck also avoids the simple static case to avoid creating the expectation that indirect code will be checked in a useful way.

**Comment by samrocketman** [comment]
How do I consider this "acceptable" behavior?  For example, I have the following code and the variable is _intended_ to expand locally before executing remotely.  I even account for the fact that it could have spaces from the local expansion when it is running remotely.

```bash
ssh -nF "${VAGRANT_SSH_CONFIG}" default "sudo test -f \"${VAGRANT_JENKINS_HOME}\"/secrets/initialAdminPassword"
```

Would it be considered a bug if I'm properly quoting the client-side expansion on the remote host to account for spaces?

**Comment by samrocketman** [comment]
@koalaman would it make sense to account for this in shellcheck.  e.g. a solution to the client side expansion could be resolved in one of two ways:
1. Fix your variable so it is not expanding locally.
2. Fix your remote logic so that it is remotely quoting the client-side expansion.

**Comment by koalaman** [comment]
It's not properly quoted here, so it could indeed be considered a bug. If your variable contains `$` or `"` or such, it'll break in spite of the literal double quotes. The correct way to quote it here (assuming both local and remote is `bash`) is:

    ssh  host "sudo test -f $(printf '%q' "${VAGRANT_JENKINS_HOME}")/secrets/initialAdminPassword"

ShellCheck doesn't recognize that this is correct though.
